osquery/tables/yara/yara.cpp

```diff
@@ -73,35 +114,82 @@ void doYARAScan(YR_RULES* rules,
+    // Check if this "ad-hoc" signature file has not been used/compiled.
+    if (rules.count(hashStr(file, YC_FILE)) == 0) {
+      // If this is a relative path append the default yara search path.
+      auto path = (file[0] != '/') ? kYARAHome : "";
+      path += file;
+
+      YR_RULES* tmp_rules = nullptr;
+      auto status = compileSingleFile(path, &tmp_rules);
+      if (!status.ok()) {
+        LOG(WARNING) << "YARA compile error: " << status.toString();
+        continue;
+      }
+      // Cache the compiled rules by setting the unique signature file path
+      // as the lookup name. Additional signature file uses will skip the
+      // compile step and be added to the scan context
+      rules[hashStr(file, YC_FILE)] = tmp_rules;
+    }
+
+    scanContext.insert(std::make_pair(YC_FILE, file));
   }
 
-  // This could be abstracted into a common "get rules for group" function.
-  auto parser = Config::getParser("yara");
-  if (parser == nullptr || parser.get() == nullptr) {
-    LOG(ERROR) << "YARA config parser plugin has no pointer";
-    return results;
+  return Status::success();
+}
+
+Status genYaraRuleFromString(QueryContext& queryContext,
+                             YaraScanContext& scanContext) {
+  auto yaraParser = getYaraParser();
+  if (yaraParser == nullptr || yaraParser.get() == nullptr) {
+    return Status::failure("YARA config parser plugin has no pointer");
   }
 
-  std::shared_ptr<YARAConfigParserPlugin> yaraParser = nullptr;
-  try {
-    yaraParser = std::dynamic_pointer_cast<YARAConfigParserPlugin>(parser);
-  } catch (const std::bad_cast& e) {
-    LOG(ERROR) << "Error casting yara config parser plugin";
-    return results;
+  auto& rules = yaraParser->rules();
+  auto sigrules = queryContext.constraints["sigrule"].getAll(EQUALS);
+
+  // Compile signature string and add them to the scan context
+  for (const auto& rule_string : sigrules) {
+    if (rules.count(hashStr(rule_string, YC_RULE)) == 0) {
+      YR_RULES* tmp_rules = nullptr;
+
+      auto status = compileFromString(rule_string, &tmp_rules);
+      if (!status.ok()) {
+        LOG(WARNING) << "YARA rule : " << rule_string << status.toString();
+        continue;
+      }
+
+      rules[hashStr(rule_string, YC_RULE)] = tmp_rules;
+    }
+    // Add rule string to the scan context set
+    scanContext.insert(std::make_pair(YC_RULE, rule_string));
   }
-  if (yaraParser == nullptr || yaraParser.get() == nullptr) {
-    LOG(ERROR) << "YARA config parser plugin has no pointer";
+
+  return Status::success();
+}
+
+QueryData genYara(QueryContext& context) {
+  QueryData results;
+
+  // Must specify a path constraint and at least one of sig_group, sigfile, or
+  // sigrule
+  auto groups = context.constraints["sig_group"].getAll(EQUALS);
+  auto sigfiles = context.constraints["sigfile"].getAll(EQUALS);
+  auto sigrules = context.constraints["sigrule"].getAll(EQUALS);
+  if (groups.empty() && sigfiles.empty() && sigrules.empty()) {
```

Comment:
Feels odd. These are checked here, but also then checked in the `genYaraRuleFrom...` routines. Do you think it's cleaner to call the `genYaraRuleFrom...` routines, and then check if `scanContext` is gt 0?

(This is not a leading question, I don't know what feels cleaner)

Reply:
I updated the checks and it will get the rules compiled and add them to the `scanContext`. It checks if  the `scanContext` is empty and return empty result.